### PR TITLE
p2p: add prototype PEX reactor for new stack

### PR DESCRIPTION
**Pull request by erikgrinaker**
This adds a prototype PEX reactor for the new P2P stack.

### CODE REVIEW

File: p2p/peer.go [hunk 1]
```diff
@@ -493,32 +493,43 @@ func (m *PeerManager) Add(address PeerAddress) error {
...
+		for _, addressInfo := range peer.AddressInfo {
+			addressEndpoints, err := addressInfo.Address.Resolve(context.Background())
+			if err != nil {
+				continue
+			}
+			for _, endpoint := range addressEndpoints {
+				if len(endpoints) >= int(limit) {
+					return endpoints
+				}
+				if endpoint.IP != nil {
```
Comment (melekes) [inline p2p/peer.go:526]: under what conditions it can be `nil`?

**Review by melekes — APPROVED** [review]

File: p2p/peer.go (hunk 1 above)
Comment (erikgrinaker) [reply p2p/peer.go:526]: For example `MemoryTransport` that we use for tests, it has addresses of the form `memory:<nodeID>`.

File: p2p/pex/reactor.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+	logger := r.Logger.With("peer", envelope.From)
+
+	// FIXME: We may want to add DoS protection here, by rate limiting peers and
+	// only processing addresses we actually requested.
+	switch msg := envelope.Message.(type) {
+	case *protop2p.PexRequest:
+		endpoints := r.peerManager.Advertise(envelope.From, maxAddresses)
+		resp := &protop2p.PexAddrs{Addrs: make([]protop2p.NetAddress, 0, len(endpoints))}
+		for _, endpoint := range endpoints {
+			// FIXME: This shouldn't rely on NetAddress.
```
Comment (alexanderbez) [inline p2p/pex/reactor.go:87]: Why should this not rely on `NetAddress`?

**Review by alexanderbez — APPROVED** [review]

File: p2p/pex/reactor.go (hunk 2 above)
Comment (erikgrinaker) [reply p2p/pex/reactor.go:87]: Because `NetAddress` is an artifact of the legacy P2P stack. The new stack uses `PeerAddress` and `Endpoint`.

**Pull request merged**